Adapt python op converters covered in paddle-3-beta2 for TRT 10

anderson101866 opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->
Inference

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Deprecations | Others ] -->
Bug fixes

### Description
<!-- Describe what you’ve done -->

- Added support for TRT 10.0 since 
   - some deprecated APIs were removed or (NOTE: we have a mapping table to help developer migrate code [here](https://docs.nvidia.com/deeplearning/tensorrt/migration-guide/index.html))
   - some interfaces are refined in TRT10


細節主要分為:
 (1) ~~[build TRT engine](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/Builder.html#tensorrt.Builder)相關的改動~~(PR審核期間，Paddle內部已移除deprecated `build_engine`, see [commit:d4e1784b](https://github.com/anderson101866/Paddle/commit/d4e1784b8caa2473247b1583b539873a7dfa971b#diff-960cee3e52e89684ae51c3727e3a120ad17749a58955693113d2129ee323b787R442)) 、以及 
 (2) [Layer](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Graph/Layers.html)相關的改動
    - 其中(1)為一次性的改動engine建置過程；而 **後續python op converter開發者僅需專注(2)** 的部分，專注TRT 10的介面變化

(1)的部分請參考`python/paddle/tensorrt/converter.py`的改動，很簡單。可以參考上文[deprecated API的表格](https://docs.nvidia.com/deeplearning/tensorrt/migration-guide/index.html)
(2)的部分包括:

-   TRT10的[IShapeLayer](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Graph/Layers.html#tensorrt.IShapeLayer)從此擴展為INT64的介面，以適應形狀大的Tensor。
    - 已知某些paddle op的輸入`shape_tensor`只有int32版本，因此TRT上/下游的paddle op實作INT64之前，只得套用現有的workaround (see: `python/paddle/tensorrt/converter_utils.py` 統一轉換成較窄的int32，直到paddle正式支援int64 shape

- TRT10對於pool layer僅支援nd的介面，移除2d介面 (同樣可參考上文的[deprecated API表格](https://docs.nvidia.com/deeplearning/tensorrt/migration-guide/index.html))
    - ~~see `python/paddle/tensorrt/impls/pooling.py`, `add_pooling`/`stride`/`padding`等介面移除~~
      然而此PR合併前，Paddle內部已於`pooling.py`中移除2d介面，see [commit:6043c7f3](https://github.com/anderson101866/Paddle/commit/6043c7f382021b2ce1c1c0a9fd8153dedb18556c#diff-da593382ac700b90bc2c41f11ba23d78307edc8db1ba7ddfb54d7f9d022a0fcd)
- 其餘改動，僅是驗證現有paddle-3.0.0-beta2有的單元測試有覆蓋的op，將支援版本限制放鬆
    - `trt_version="8.x"` -> `trt_version="trt_version_ge=8.0"`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/Paddle?pullRequest=69510) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/PaddlePaddle/Paddle?pullRequest=69510) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/PaddlePaddle/Paddle?pullRequest=69510) before we can accept your contribution.<br/><hr/>**Anderson Meng** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/PaddlePaddle/Paddle?pullRequest=69510) it.</sub>

anderson101866 committed:
Adapt `paddle.tensorrt.converter` to TRT 10

This commit includes basic migration to TRT 10 API, and also enable
 those converter who are tested with existing unittest to TRT10

For those 2 converter which is NOT included in this commit:
- python/paddle/tensorrt/impls/attribute.py
- python/paddle/tensorrt/impls/common.py
Need to fix these 2 behavior issues in later commit.

anderson101866 committed:
Adapt "pd_op.shape" TRT op converter to TRT10

Although `IShapeLayer` supports shape in int64 since TRT10, some paddle
 native op kernel only implements their input shape tensor (if exists)
 in int32. Hence, there is a workaround in `trt_shape` to cast the
 result of TRT `IShapeLayer` back to int32 to be more compatible with
 other paddle op. (see python/paddle/tensorrt/converter_utils.py)

Please remove the workaround when all paddle op supports their shape in
 int64.

Also, since `IShapeLayer` return shape in int64 in TRT10, the
 "pd_op.shape64" will be seamlessly supported in TRT10 w/o any extra
  workaround.

anderson101866 committed:
Fix converter error in TRT10 for interpolation ops

Error detail:
  {
      (%1) = "pd_op.bilinear_interp" [id:28] (%2, %3, <<NULL VALUE>>, <<NULL VALUE>>) {__l_trt__:true,align_corners:false,align_mode:(Int32)0,data_format:"NCHW",interp_method:"bilinear",out_d:(Int32)-1,out_h:(Int32)12,out_w:(Int32)12,scale:[],stop_gradient:[true]} : (builtin.tensor<-1x3x6x10xf32>, builtin.tensor<2xi32>, <<NULL TYPE>>, <<NULL TYPE>>) -> builtin.tensor<-1x3x12x12xf32>
      () = "cf.yield" [id:36] (%1) {} : (builtin.tensor<-1x3x12x12xf32>) ->
  }
  [TRT] [E] ITensor::getDimensions: Error Code 4: API Usage Error ((Unnamed Layer* 6) [Concatenation]: concat input tensors 0 and 2 have incompatible types Int64 and Int32)
  [TRT] [E] IBuilder::buildSerializedNetwork: Error Code 4: API Usage Error ((Unnamed Layer* 6) [Concatenation]: concat input tensors 0 and 2 have incompatible types Int64 and Int32)

The error happened in "python/paddle/tensorrt/impls/common.py" because
 IConcatenationLayer requires all input to be same dtype.
The (shape)tensor passed from paddle op will be int32, while the TRT
 IShapeLayer supports int64 shape; thereby, cannot be concatenated with
 each other.

Here, we call `trt_shape` to get shape tensor whose dtype aligned with
 the dtype from paddle op.

anderson101866 committed:
Add int64 in TRT->paddle dtype mapping function

anderson101866 committed:
Fix "test_converter_math" by enable op for TRT 10

test_converter_math can pass unittest with environment with TRT 10.6

anderson101866 committed:
Adapt 3 manipulation converters to TRT 10

- "pd_op.expand"
- "pd_op.expand_as"
- "pd_op.slice"

YuanRisheng commented:
目前CI流水线上缺乏对TRT10版本的监控，PR先暂时这样提上来，可能先不合入，等最近协调后应该会上线带有TRT10的流水线，到时候再合入PR，谢谢

anderson101866 commented:
@YuanRisheng 
了解了解，另外只是確認一下工作流程:
因為現在TRT10的python converter flow **完全不能** 運行，所以即使有TRT10的CI流程，CI保護的flow本身 **現況** 就是壞的，所有python converter單測現況都會failure。這是預期的對嗎?

YuanRisheng commented:
> @YuanRisheng 了解了解，另外只是確認一下工作流程:
> 
> 1. 因為現在TRT10的python converter flow **完全不能** 運行，所以即使有TRT10的CI流程，CI保護的flow本身 **現況** 就是壞的，所有python converter單測現況都會failure。這是預期的對嗎?
> 2. 另外，
>    [PR-CI-APPROVAL](https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/detail/11913735/job/28064465)當中回報:
>    > 0. You must have one RD (From00, zhangbo9674) approval for file changes in paddle/fluid/framework/new_executor.
>    
>    
>    屆時會有人幫我approve這底層fluid相關的改動，對吧?

@anderson101866 对于第一点你的理解应该是对的，这个我们再考虑一下， 想一想是先让你们这边把修复trt10的单测pr提交合入，还是先把trt10的ci弄起来（convert相关单测禁止运行），你们这边提交修复代码后打开单测。对于第二点，你的理解也是对的，我们这边会找人来让CI-APPROVAL通过

anderson101866 commented:
好的謝謝，另外附註，NV這邊因為環境關係，我們只有beta2的環境，所以只有打開幾個beta2單測有覆蓋的op。

實際develop分支的op數量應該更多，這個PR未包含

anderson101866 commented:
Is the CI ready?  or it'll continuously suffer conflict from newly submitted change.

anderson101866 commented:
已更新PR:

- pool 2d deprecated API在此PR審核期間，已被同仁[commit](https://github.com/anderson101866/Paddle/commit/6043c7f382021b2ce1c1c0a9fd8153dedb18556c#diff-da593382ac700b90bc2c41f11ba23d78307edc8db1ba7ddfb54d7f9d022a0fcd)內部移除。故此PR捨棄對應的change
- build_engine此TRT 8 API，已deprecated，也已被同仁[commit](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/Builder.html#tensorrt.Builder)內部移除，此PR同樣捨棄對應的change

YuanRisheng commented:
CI-Codestyle-Check存在代码格式的问题，请本地修正后再提交一次，谢谢

anderson101866 commented:
OK重解conflict的時候沒有注意到，已上傳 等待check重新驗證

anderson101866 commented:
@YuanRisheng YuanRisheng [PR-CI-APPROVAL](https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/detail/12016152/stage/18075184/0?jump=latestJob) 這一項，貌似牽涉到paddle底層fluid，需要特別的人review approve，請在幫忙一下 謝謝

vivienfanghuagood reviewed (approved):
LGTM
